### PR TITLE
Add `dimension_names` array metadata field

### PR DESCRIPTION
**Pull request by jbms**

### CODE REVIEW

**Comment by jstriebel** [comment]
Fixes #73

File: docs/core/v3.0.rst
Comment (shoyer) [inline docs/core/v3.0.rst:1099]: Why not use `null` instead for unnamed dimensions?

File: docs/core/v3.0.rst
Comment (shoyer) [inline docs/core/v3.0.rst:1100]: I'm not sure how I feel about the uniqueness restriction. I would be inclined to omit it, and leave it up to applications/extensions to enforce.

Comment (jbms) [reply docs/core/v3.0.rst:1100]: Do you have in mind a use case for duplicate dimension names, or do you know of any software that could make use of that?

Comment (shoyer) [reply docs/core/v3.0.rst:1100]: Duplicate dimension names come up for cases like covariance matrices. In Xarray, we support creating them (I have mixed feelings about this), but you can't do named operations in the case of duplicates.

File: docs/core/v3.0.rst
Comment (shoyer) [reply docs/core/v3.0.rst:1099]: I would prefer enforcing this at the type level, e.g., "dimension names are any valid string."

Otherwise, I imagine _somebody_ is going to try to set dimension names to an empty string in their application code and wonder why they don't round-trip to Zarr properly.

File: docs/core/v3.0.rst
Comment (jbms) [reply docs/core/v3.0.rst:1100]: Thanks for that clarification --- I didn't think about that use case.  I guess in xarray the dimension names serve (at least) two different purposes: associating coordinate arrays and allowing dimensions to be referenced by name, and duplicate dimension names still are useful for specifying coordinate arrays.  Does netcdf also allow duplicate dimension names?

In tensorstore dimension names are solely used for allowing dimensions to be referenced by name, and duplicate names are not currently supported.

File: docs/core/v3.0.rst
Comment (jbms) [reply docs/core/v3.0.rst:1099]: The current proposal will allow any string to round trip ---- the uniqueness constraint just applies to non-empty strings.  Are you proposing that empty strings should be supported and be treated exactly the same as  non-empty strings --- and that unnamed dimensions should be identified separately using null?

Comment (shoyer) [reply docs/core/v3.0.rst:1099]: > Are you proposing that empty strings should be supported and be treated exactly the same as non-empty strings --- and that unnamed dimensions should be identified separately using null?

Yes. Otherwise application code will not be able to make the distinction (if desired) between "unnamed" and "named with an empty string."

File: docs/core/v3.0.rst
Comment (shoyer) [reply docs/core/v3.0.rst:1100]: Yes, NetCDF does allow for duplicate dimension names.

Comment (ivirshup) [reply docs/core/v3.0.rst:1100]: * Cross referencing some relevant xarray issue: https://github.com/pydata/xarray/issues/3731

Another use case here is storing graphs as adjacency matrices. In AnnData (but also other projects like pytorch-geometric) we have an set of observations (nodes) with features in tables and tensors,  aligned with graphs represented as adjacency matrices. If we want to subset or sort the observations, it's important that these operations are applied to both dimensions of the adjacency matrix.

This use case has been coming up a lot in discussions around AnnData's integration into OME-NGFF, where you'd have a graph connecting neighboring regions.

Comment (jbms) [reply docs/core/v3.0.rst:1100]: Do NetCDF implementations support indexing operations using dimension names to specify dimensions?  If so, how do they handle duplicate dimension names?  Is there any software that actually provides the desired behavior, where if you perform an indexing operation using a named dimension, and there is more than one dimension with the given name, then the indexing operation applies to all of the dimensions?

It seems to me that duplicate dimension names just don't work very well --- if you have duplicate dimension names, then even if there are some indexing operations that should be applied uniformly to all of the identically-named dimensions, there will almost surely be other indexing operations that should be applied to only one or some of the identically-named dimensions --- for example, to look up a non-diagonal entry in the covariance matrix or adjacency matrix.  Since there is no way to uniquely identify such dimensions by name, they have to be specified by dimension index, and you therefore lose the benefit of dimension names.

A better solution might be to keep dimension names unique, but add support for a special syntax to xarray to allow the dimension name not to exactly equal to the coordinate array name.  For example, the coordinate array might be named "x" and the dimension names might be "x:1" and "x:2" (just to toss out an arbitrary syntax that I haven't given much thought to).  xarray could still allow indexing operations on "x" to apply to both dimensions.

**Comment by jstriebel** [comment]
Crosslinking https://github.com/zarr-developers/zarr-specs/pull/149#discussion_r927300522

File: docs/core/v3.0.rst
Comment (shoyer) [reply docs/core/v3.0.rst:1100]: > Do NetCDF implementations support indexing operations using dimension names to specify dimensions?

The official netCDF-C library does not. Dimension names are just kept around as metadata.

> It seems to me that duplicate dimension names just don't work very well

I agree. Certainly the semantics of Xarray's indexing with labeled dimensions break with duplicates.

That said, I don't think "not working well" in certain data models is a good reason for making the Zarr spec more complex to prohibit certain behavior. There are clearly cases that could justify storing arrays with duplicate dimensions names (e.g., covariance and adjacency matrices), especially if not using APIs that look-up dimensions by name.

Xarray and NetCDF do impose some additional restrictions on dimension names, namely:

- No missing dimension names
- The size of array axes corresponding to a dimension name must be consistent between all arrays in a group

These restrictions make sense for Xarray/NetCDF, but I would not impose them for all uses of `dimension_names` in Zarr. Instead, at most I would add a comment like "Extensions may impose additional restrictions on valid dimension names" with a link to the NetCDF Zarr spec (when complete).

> A better solution might be to keep dimension names unique, but add support for a special syntax to xarray to allow the dimension name not to exactly equal to the coordinate array name.

I agree, this sort of thing could make sense for Xarray.

Comment (jbms) [reply docs/core/v3.0.rst]: Per your comment, I changed the requirement to be a recommendation.

Comment (ivirshup) [reply docs/core/v3.0.rst]: > Is there any software that actually provides the desired behavior, where if you perform an indexing operation using a named dimension

AnnData sorta does this, but it's a very constrained data model where we don't really need to do matching dimensions by name.

------

I think the current language resolves this for me.

**Review by rabernat — APPROVED** [review]

**Review by jstriebel — DISMISSED** [review]
LGTM :+1:

**Review by jstriebel — DISMISSED** [review]

**Comment by jstriebel** [comment]
@rabernat I just resolved a merge conflict which dismissed our old reviews, and mine doesn't count anymore. Could you re-approve this one please? Thanks :-)

**Comment by shoyer** [comment]
I don't mean to hold things up with bike-shedding, but did we reach concensus on using `""` vs `null` for missing dimension names?

**Comment by jbms** [comment]
> I don't mean to hold things up with bike-shedding, but did we reach concensus on using `""` vs `null` for missing dimension names?

Is your intention to allow the empty string as a valid dimension name distinct from null, and not meaning an unlabeled dimension?

**Comment by shoyer** [comment]
> > I don't mean to hold things up with bike-shedding, but did we reach concensus on using `""` vs `null` for missing dimension names?
> 
> Is your intention to allow the empty string as a valid dimension name distinct from null, and not meaning an unlabeled dimension?

My thinking was that `null` instead of a string would indicate missing dimension names, with no special meaning for empty strings.

I prefer `null` because it's less ambiguous from a typing perspective -- if somebody writes code that doesn't anticipate partially missing dimension names, it will fail noisily rather than silently.

**Comment by jstriebel** [comment]
@jbms As discussed last week, I changed this here to use null objects for unnamed dimensions, as requested by @shoyer. I believe this is ready to be merged now. My approval does not count anymore since I did some changes, @joshmoore could you approve this one?

**Review by rabernat — APPROVED** [review]

**Comment by rabernat** [comment]
I got you @jstriebel! 🚀

**Pull request merged**
